### PR TITLE
[core][autoscaler][v2] node providers - kuberay cloud provider[3/x] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR refactors the existing kuberay node providers and add the new kuberay cloud provider that's compatible with the `ICloudInstanceProvider` interface for autoscaler v2. 

The main bulk of logic should be the same, with some refactoring and changes to data structures and code. 

**Major difference of v1/v2 provider:**
The v2 provider took away the `post_process` logic as it no longer assumes the caller would make a sequence of calls to 
`non_terminated_nodes()` -> `launch` -> `terminate` -> `post_process` in a loop. Instead, v2 cloud provider will check any pending to deletes workers for `launch` and `terminate` independently. In the event of with pending workers to delete, the v1 and v2 should behave the same (no updates to the CR) 

TODO:
- Add more e2e testing for kuberay once other parts of autoscaler v2 are code-complete. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/kuberay/node_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/_private/kuberay/node_provider.py:208]: i can potentially move these changes to another PR.

Comment (rickyyx) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:208]: this is merely moving v1's codes around into a client class that's mockable for v2 provider testing.

File: python/ray/autoscaler/_private/kuberay/node_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/_private/kuberay/node_provider.py:294]: v1 provider would still own the client w/o having to change caller's codes.

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:258]: This is largely adopted from https://github.com/ray-project/ray/blob/67a5b717b5e601543d3afd41387e5c79ebe30b7f/python/ray/autoscaler/_private/kuberay/node_provider.py#L434-L469

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:309]: However, different from the original impl, I added the clear worker groups logic and no longer have the issue of side-effect mentioned here: https://github.com/ray-project/ray/blob/67a5b717b5e601543d3afd41387e5c79ebe30b7f/python/ray/autoscaler/_private/kuberay/node_provider.py#L372-L388

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:430]: mainly from https://github.com/ray-project/ray/blob/67a5b717b5e601543d3afd41387e5c79ebe30b7f/python/ray/autoscaler/_private/kuberay/node_provider.py#L306-L352

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:475]: This is the only change from v1's impl.

**Comment by rickyyx** [comment]
also cc @DmitriGekhtman if you have some feedbacks on this, appreciate it.

File: python/ray/autoscaler/_private/kuberay/node_provider.py
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:208]: It's a nice refactor!

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (DmitriGekhtman) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:127]: Waiting for workers to be created before deleting them?
What does this mean?

File: python/ray/autoscaler/v2/instance_manager/node_provider.py
Comment (DmitriGekhtman) [inline python/ray/autoscaler/v2/instance_manager/node_provider.py:45]:
```suggestion
    # associate the cloud instance with an update request like KubeRay.
```

**Review by DmitriGekhtman — COMMENTED** [review]
looks pretty reasonable to me :)

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:127]: Oops, types. Should be "waiting for them to be deleted before adding new workers to be delete"

**Comment by rickyyx** [comment]
> looks pretty reasonable to me :)

Thanks for the prompt feedback!

File: python/ray/autoscaler/_private/kuberay/node_provider.py
Comment (rickyyx) [inline python/ray/autoscaler/_private/kuberay/node_provider.py:209]: maybe this should be called `IKubernetesApiClient`

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:115]: Do all the requests, no matter `terminate` or `launch`, have a unique ID?

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:38]:
```suggestion
class KubeRayProvider(ICloudInstanceProvider):
```

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:426]: This is weird. All worker groups belonging to `worker_groups_with_pending_deletes` will also be added to `worker_groups_with_finished_deletes`. These two sets should be mutually exclusive.

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:222]:
```suggestion
        # Add to launch nodes.
```

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:226]: The two loops `for to_delete_id in to_delete_instances` and `for instance_id in to_delete_instances` are very similar. Can we fuse them? If we can't, we should at least have a consistent naming convention (`to_delete_id`, `instance_id`).

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:252]: Can we rename `worker_groups_to_clear_delete` to `worker_groups_without_pending_deletes`?

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:114]: Just for my education, who calls the `terminate` and `launch` functions?

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:173]: Is this only for testing?

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:261]:
```suggestion
        """Submits a scale request to the Kubernetes API server.
```

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:264]:
```suggestion
        to the Kubernetes API server.
```

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (kevin85421) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:270]:
```suggestion
            Exception: An exception is raised if the Kubernetes API server returns an
```

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:114]: instance manager (and its subscribers)

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:115]: Yes.

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:426]: oh good catch! i wonder how come it wasn't caught with my test.

Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:426]: hmm, i think this doesn't affect correctness since if any worker groups with pending deletes, we will not scale. So when there's no worker groups with pending deletes, all the worker groups will have finished deletes.

But yeah, this could be made mutually exclusive.

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:460]: looks like `fetched_resource_version` was merely logged. I wasn't sure why we used it in the first place. cc @DmitriGekhtman

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:487]: oh good to know!

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:510]: yes - i believe a well defined structure is better, I feel this is another reason I feel use KubeRay operator sits between k8s API server and node provider is better since all the details of pod schema could be hidden by KubeRay operator.

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:512]: i see, good to learn!

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:515]: running here doesn't have much meaning for ray. All it needs to know is pending + running actually. As long as it's pending + running, it's considered an allocated cloud instance.

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:173]: it's part of the interface for `CloudInstanceProvider`

**Comment by DmitriGekhtman** [comment]
resource version part is a bit subtle
I'm just leaving this comment as a mark to remind myself to come back and explain a bit more in detail

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (DmitriGekhtman) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:460]: The `fetched_resource_version` is at least the supplied `resource_version` in the list request.
It is logged simply for observability to reasons, to give an indication of the potential staleness of the returned pod list data.

Although the K8s API forbids making decisions based on resource version comparisons, it could be useful (at least for maintainers of this code, at least for the ones who wrote this code...) to have a sense of whether this fetched resource version is _at least increasing_ across autoscaler operations.

Feel free to drop the bit of code around the fetched_resource_version if you wish; I essentially added the log line _for myself_.

Providing a bit more context:
The resource version is included in the request with a "not older than parameter" to improve scalability (without this parameter, the requests could be harmful to the health of the K8s cluster, especially when many Ray clusters are running).
Previously, I tried to extract a _recent_ resource version by writing a patch to head pod and then reading the head pod object back; however users complained that the illicit write to the head pod was screwing with their CI/CD setups. So we no longer write to the head pod and just keep reusing the resource version at which the head pod was last updated (the head pod was potentially last updated when it started running), just accept that the read might be a little stale, and hope that the read is at least fresh enough to reflect changes since the last autoscaler update.
In theory, **this is sketch**. In practice, it seems that this hasn't caused any issues for anyone...yet...that we've heard of.

**Comment by DmitriGekhtman** [comment]
> resource version part is a bit subtle
I'm just leaving this comment as a mark to remind myself to come back and explain a bit more in detail

https://github.com/ray-project/ray/pull/42720#discussion_r1475447095

**Comment by scv119** [comment]
better leave this to @kevin85421 and @DmitriGekhtman to review

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:460]: Thanks for the info! This is very helpful for learning the context. 

Doesn't look like we are using `fetched_resource_version` other than logging? I am fine with keeping things as it's now and we could revisit it.

**Comment by rickyyx** [comment]
@kevin85421 @DmitriGekhtman @jjyao I think I addressed most (if not all) the comments and nits, lmk what you folks think.

**Comment by DmitriGekhtman** [comment]
Looked reasonable when I last looked at it.
Unfortunately, I don't have bandwidth for a careful review.

The key thing is to test thoroughly.

**Comment by rickyyx** [comment]
> Looked reasonable when I last looked at it. Unfortunately, I don't have bandwidth for a careful review.
> 
> The key thing is to test thoroughly before rolling out to KubeRay users.

Yes - will at least port all the existing kuberay/ray tests available in Ray for test.

**Review by kevin85421 — APPROVED** [review]
Overall, LGTM. This PR is too big to review in detail. Let's make v2 end-to-end work as soon as possible with KubeRay so that we can add more e2e tests.

**Review by jjyao — APPROVED** [review]

**Pull request merged**
